File: src/main/scala/org/tindalos/principle/infrastructure/reporters/packagestructure/ExistingPackageCohesionsFileWriter.scala
package org.tindalos.principle.infrastructure.reporters.packagestructure

import java.io.PrintWriter

import org.tindalos.principle.domain.analyzers.structure.CohesionAnalysisResult
import org.tindalos.principle.domain.analyzers.structure.NodeGroup
import scala.collection.JavaConverters._
import org.tindalos.principle.infrastructure.reporters.ReportsDirectoryManager
import org.tindalos.principle.infrastructure.reporters.packagestructure.PackageCohesionAnalysisResultReporter.{packageCohesionsFileName, generalDescription, round}
import PackageCohesionConstants._

object ExistingPackageCohesionsFileWriter {

  def writeToFile(result: CohesionAnalysisResult) = {

    val columns = "| Cohesion | Size | Upstream/Downstream dependencies of package | Internal/External edges of the classes |\n"
    val oneElementPackageDesc = "Each of the following packages contains only one class, therefore no cohesion is calculated"

    val printWriter = new PrintWriter(ReportsDirectoryManager.reportDirectoryPath+"/"+packageCohesionsFileName)

    printWriter
      .append(generalDescription)
      .append("\n\n" + SECTION_LINE + "\n")
      .append("Package cohesions - existing packages are listed, ordered by cohesion")
      .append("\n" + SECTION_LINE + "\n\n")
      .append(columns)

    result.packages()
      .entrySet().asScala
      .filter(_.getValue.nodes.size > 1)
      .toList
      .sortBy(_.getValue.cohesion()).reverse
      .foreach {
      e => printWriter.append(s"\n ${groupToLine(e.getValue)}\t${e.getKey}")
    }

    printWriter.append("\n"+SECTION_LINE+"\n")
    val minPackages = result.packages().entrySet().asScala.filter(_.getValue.nodes.size == 1)
    printWriter.append(s"\n${oneElementPackageDesc} (${minPackages.size})\n\n")
    minPackages.toList.map(_.getValue.nodes.asScala.head.id).sorted.foreach {
      nodeId => printWriter.append( nodeId + "\n")
    }

    printWriter.close()
  }

  private def groupToLine(n: NodeGroup) =
    s"${round(n.cohesion())}\t" +
      s"| ${n.nodes.size}\t" +
      s"| ${n.externalDependants.size}/${n.externalDependencies.size}\t" +
      s"| ${n.internalArcsNo}/${n.externalArcsNo} | "

}

File: src/main/java/org/tindalos/guardrails/internal/infrastructure/reporters/packagestructure/ExistingPackageCohesionsFileWriter.java
package org.tindalos.guardrails.internal.infrastructure.reporters.packagestructure;

import java.io.FileNotFoundException;
import java.io.PrintWriter;
import java.util.Comparator;
import java.util.Map;

import org.tindalos.guardrails.internal.domain.analyzers.structure.CohesionAnalysisResult;
import org.tindalos.guardrails.internal.domain.analyzers.structure.NodeGroup;
import org.tindalos.guardrails.internal.infrastructure.reporters.ReportsDirectoryManager;
import static org.tindalos.guardrails.internal.infrastructure.reporters.packagestructure.PackageCohesionConstants.GENERAL_DESCRIPTION;
import static org.tindalos.guardrails.internal.infrastructure.reporters.packagestructure.PackageCohesionConstants.PACKAGE_COHESIONS_FILE_NAME;
import static org.tindalos.guardrails.internal.infrastructure.reporters.packagestructure.PackageCohesionConstants.SECTION_LINE;
import static org.tindalos.guardrails.internal.infrastructure.reporters.packagestructure.PackageCohesionConstants.round;

/**
 * Writes package cohesion analysis results for existing packages to a plain text report.
 */
final class ExistingPackageCohesionsFileWriter {

    private static final String COLUMNS = """
            | Cohesion | Size | Upstream/Downstream dependencies of package | Internal/External edges of the classes |
            """;
    private static final String ONE_ELEMENT_PACKAGE_DESCRIPTION =
            "Each of the following packages contains only one class, therefore no cohesion is calculated";
    private static final String REPORT_HEADER = """

            %s
            Package cohesions - existing packages are listed, ordered by cohesion
            %s

            """;

    private ExistingPackageCohesionsFileWriter() {
    }

        static void writeToFile(CohesionAnalysisResult result) {
        try (var printWriter = new PrintWriter(ReportsDirectoryManager.reportDirectoryPath + "/" + PACKAGE_COHESIONS_FILE_NAME)) {
            printWriter
                    .append(GENERAL_DESCRIPTION)
                    .append(REPORT_HEADER.formatted(SECTION_LINE, SECTION_LINE))
                    .append(COLUMNS);

            result.packages().entrySet().stream()
                    .filter(entry -> entry.getValue().nodes().size() > 1)
                    .sorted(Comparator.comparingDouble((Map.Entry<String, NodeGroup> entry) -> entry.getValue().cohesion()).reversed())
                    .forEach(entry -> printWriter.append("\n " + groupToLine(entry.getValue()) + "\t" + entry.getKey()));

            printWriter.append("\n" + SECTION_LINE + "\n");
            var singleClassPackages = result.packages().entrySet().stream()
                    .filter(entry -> entry.getValue().nodes().size() == 1)
                    .toList();
            printWriter.append("""

                    %s (%d)

                    """.formatted(ONE_ELEMENT_PACKAGE_DESCRIPTION, singleClassPackages.size()));
            singleClassPackages.stream()
                    .map(entry -> entry.getValue().nodes().iterator().next().id())
                    .sorted()
                    .forEach(nodeId -> printWriter.append(nodeId).append("\n"));
        } catch (FileNotFoundException exception) {
            throw new RuntimeException("Failed to write existing package cohesions report to " +
                    ReportsDirectoryManager.reportDirectoryPath + "/" + PACKAGE_COHESIONS_FILE_NAME, exception);
        }
    }

    private static String groupToLine(NodeGroup nodeGroup) {
        return """
                %s	| %d	| %d/%d	| %d/%d |
                """
                .formatted(
                        round(nodeGroup.cohesion()),
                        nodeGroup.nodes().size(),
                        nodeGroup.externalDependants.size(),
                        nodeGroup.externalDependencies.size(),
                        nodeGroup.internalArcsNo,
                        nodeGroup.externalArcsNo
                )
                .stripTrailing();
    }
}